provisioner/converge/provisioner.go

```diff
@@ -0,0 +1,240 @@
+				Exclude: []string{"execute_command"},
+			},
+		},
+		raws...,
+	)
+	if err != nil {
+		return err
+	}
+
+	// require a single module
+	if p.config.Module == "" {
+		return errors.New("Converge requires a module to provision the system")
+	}
+
+	// set defaults
+	if p.config.WorkingDirectory == "" {
+		p.config.WorkingDirectory = "/tmp"
+	}
+
+	if p.config.ExecuteCommand == "" {
+		p.config.ExecuteCommand = "cd {{.WorkingDirectory}} && sudo converge apply --local --log-level=WARNING --paramsJSON '{{.ParamsJSON}}' {{.Module}}"
+	}
+
+	// validate version
+	if !versionRegex.Match([]byte(p.config.Version)) {
+		return fmt.Errorf("Invalid Converge version %q specified. Valid versions include only letters, numbers, dots, and dashes", p.config.Version)
+	}
+
+	// validate sources and destinations
+	for i, dir := range p.config.ModuleDirs {
+		if dir.Source == "" {
+			return fmt.Errorf("Source (\"source\" key) is required in Converge module dir #%d", i)
+		}
+		if dir.Destination == "" {
+			return fmt.Errorf("Destination (\"destination\" key) is required in Converge module dir #%d", i)
+		}
+	}
+
+	return err
+}
+
+// Provision node somehow. TODO: actual docs
+func (p *Provisioner) Provision(ui packer.Ui, comm packer.Communicator) error {
+	ui.Say("Provisioning with Converge")
+
+	// bootstrapping
+	if err := p.maybeBootstrap(ui, comm); err != nil {
+		return err // error messages are already user-friendly
+	}
+
+	// send module directories to the remote host
+	if err := p.sendModuleDirectories(ui, comm); err != nil {
+		return err // error messages are already user-friendly
+	}
+
+	// apply all the modules
+	if err := p.applyModules(ui, comm); err != nil {
+		return err // error messages are already user-friendly
+	}
+
+	return nil
+}
+
+func (p *Provisioner) maybeBootstrap(ui packer.Ui, comm packer.Communicator) error {
+	if !p.config.Bootstrap {
+		return nil
+	}
+	ui.Message("bootstrapping converge")
+
+	bootstrap, err := http.Get("https://get.converge.sh")
+	if err != nil {
+		return fmt.Errorf("Error downloading bootstrap script: %s", err) // TODO: is github.com/pkg/error allowed?
+	}
+	if err := comm.Upload("/tmp/install-converge.sh", bootstrap.Body, nil); err != nil {
+		return fmt.Errorf("Error uploading script: %s", err)
+	}
+	if err := bootstrap.Body.Close(); err != nil {
+		return fmt.Errorf("Error getting bootstrap script: %s", err)
+	}
+
+	// construct command
+	command := "/bin/sh /tmp/install-converge.sh"
+	if p.config.Version != "" {
+		command += " -v " + p.config.Version
+	}
+
+	var out bytes.Buffer
+	cmd := &packer.RemoteCmd{
+		Command: command,
+		Stdin:   nil,
+		Stdout:  &out,
+		Stderr:  &out,
+	}
+
+	if err = comm.Start(cmd); err != nil {
+		return fmt.Errorf("Error bootstrapping converge: %s", err)
+	}
+
+	cmd.Wait()
+	if cmd.ExitStatus != 0 {
+		ui.Error(out.String())
+		return errors.New("Error bootstrapping converge")
+	}
+
+	ui.Message(strings.TrimSpace(out.String()))
+	return nil
```

Comment:
This is a bit of an artifact… I thought "surely there must be situations where internet access is not to be had." But the script that it's running pulls from the internet, so I'll change it. :man_shrugging: